File: app/src/main/java/wangdaye/com/geometricweather/weather/apis/MfWeatherApi.java
package wangdaye.com.geometricweather.weather.apis;

import java.util.List;

import io.reactivex.rxjava3.core.Observable;
import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.Query;
import wangdaye.com.geometricweather.weather.json.mf.MfCurrentResult;
import wangdaye.com.geometricweather.weather.json.mf.MfEphemerisResult;
import wangdaye.com.geometricweather.weather.json.mf.MfForecastV2Result;
import wangdaye.com.geometricweather.weather.json.mf.MfLocationResult;
import wangdaye.com.geometricweather.weather.json.mf.MfRainResult;
import wangdaye.com.geometricweather.weather.json.mf.MfWarningsResult;

/**
 * API Météo France
 */

public interface MfWeatherApi {

    @GET("places")
    Call<List<MfLocationResult>> callWeatherLocation(@Header("User-Agent") String userAgent,
                                                     @Query("q") String q,
                                                     @Query("lat") double lat,
                                                     @Query("lon") double lon,
                                                     @Query("token") String token);

    @GET("places")
    Observable<List<MfLocationResult>> getWeatherLocation(@Header("User-Agent") String userAgent,
                                                          @Query("q") String q,
                                                          @Query("lat") double lat,
                                                          @Query("lon") double lon,
                                                          @Query("token") String token);

    @GET("v2/forecast")
    Observable<MfForecastV2Result> getForecastV2(@Header("User-Agent") String userAgent,
                                                 @Query("lat") double lat,
                                                 @Query("lon") double lon,
                                                 @Query("formatDate") String formatDate,
                                                 @Query(encoded = true, value = "instants") String instants,
                                                 @Query("token") String token);

    @GET("v2/observation")
    Observable<MfCurrentResult> getCurrent(@Header("User-Agent") String userAgent,
                                           @Query("lat") double lat,
                                           @Query("lon") double lon,
                                           @Query("lang") String lang,
                                           @Query("formatDate") String formatDate,
                                           @Query("token") String token);

    @GET("v3/nowcast/rain")
    Observable<MfRainResult> getRain(@Header("User-Agent") String userAgent,
                                     @Query("lat") double lat,
                                     @Query("lon") double lon,
                                     @Query("lang") String lang,
                                     @Query("formatDate") String formatDate,
                                     @Query("token") String token);

    @GET("ephemeris")
    Observable<MfEphemerisResult> getEphemeris(@Header("User-Agent") String userAgent,
                                               @Query("lat") double lat,
                                               @Query("lon") double lon,
                                               @Query("lang") String lang,
                                               @Query("formatDate") String formatDate,
                                               @Query("token") String token);

    @GET("v2/warning/full")
    Observable<MfWarningsResult> getWarnings(@Header("User-Agent") String userAgent,
                                             @Query(encoded = true, value = "domain") String domain,
                                             @Query("formatDate") String formatDate,
                                             @Query("token") String token);
}

File: app/src/main/java/wangdaye/com/geometricweather/weather/apis/MfWeatherApi.kt
package wangdaye.com.geometricweather.weather.apis

import retrofit2.http.GET
import retrofit2.http.Query
import wangdaye.com.geometricweather.weather.json.mf.*

/**
 * API Météo France
 */
interface MfWeatherApi {

    @GET("places")
    suspend fun callWeatherLocation(
            @Query("q") q: String,
            @Query("lat") lat: Double,
            @Query("lon") lon: Double,
            @Query("token") token: String
    ): List<MfLocationResult>

    @GET("forecast")
    suspend fun getForecast(
            @Query("lat") lat: Double,
            @Query("lon") lon: Double,
            @Query("lang") lang: String,
            @Query("token") token: String
    ): MfForecastResult

    @GET("v2/forecast")
    suspend fun getForecastV2(
            @Query("lat") lat: Double,
            @Query("lon") lon: Double,
            @Query("lang") lang: String,
            @Query("token") token: String
    ): MfForecastV2Result

    @GET("forecast")
    suspend fun getForecastInstants(
            @Query("lat") lat: Double,
            @Query("lon") lon: Double,
            @Query("lang") lang: String,
            @Query("instants") instants: String,
            @Query("token") token: String
    ): MfForecastResult

    @GET("forecast")
    suspend fun getForecastInseepp(
            @Query("id") id: Int,
            @Query("lang") lang: String,
            @Query("token") token: String
    ): MfForecastResult

    @GET("observation/gridded")
    suspend fun getCurrent(
            @Query("lat") lat: Double,
            @Query("lon") lon: Double,
            @Query("lang") lang: String,
            @Query("token") token: String
    ): MfCurrentResult

    @GET("rain")
    suspend fun getRain(
            @Query("lat") lat: Double,
            @Query("lon") lon: Double,
            @Query("lang") lang: String,
            @Query("token") token: String
    ): MfRainResult

    @GET("ephemeris")
    suspend fun getEphemeris(
            @Query("lat") lat: Double,
            @Query("lon") lon: Double,
            @Query("lang") lang: String,
            @Query("token") token: String
    ): MfEphemerisResult

    @GET("warning/full")
    suspend fun getWarnings(
            @Query("domain") domain: String,
            @Query("formatDate") formatDate: String?,
            @Query("token") token: String
    ): MfWarningsResult
}